sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -1227,6 +1227,49 @@ case class Pivot(
   override protected def withNewChildInternal(newChild: LogicalPlan): Pivot = copy(child = newChild)
 }
 
+/**
+ * A constructor for creating a melt, which will later be converted to a [[Expand]]
+ * during the query analysis.
+ *
+ * An empty values array will be replaced during analysis with all resolved outputs of child except
+ * the ids. This expansion allows to easily melt all non-id columns.
+ *
+ * The type of the value column is derived from all value columns during analysis once all values
+ * are resolved. All values' types have to be compatible, otherwise the result value column cannot
+ * be assigned the individual values and an AnalysisException is thrown.
+ *
+ * @see `org.apache.spark.sql.catalyst.analysis.TypeCoercionBase.MeltCoercion`
+ *
+ * @param ids                Id columns
+ * @param values             Value columns to melt
+ * @param variableColumnName Name of the variable column
+ * @param valueColumnName    Name of the value column
+ * @param valueType          Type of value column once known
+ * @param child              Child operator
+ */
+case class Melt(
+    ids: Seq[NamedExpression],
+    values: Seq[NamedExpression],
```

Comment:
I think id and value columns can simply be `Expression` instead of `NamedExpression`. In the analyzer rule, we can fail if the id and value columns end up with normal expressions, e.g. `df.melt(Array($"id" + 1), ...)`

Reply:
The `values` have to be `NamedExpression`, as I need the name for the variable column. Or how else could I get a name from an `Expression`?

And how could I turn an Expression into an Attribute (needed for `output` in `Expand`)? Pointers or examples appreciated.

And why should normal expressions not be allowed for `ids` and `values`? Allows for more expressive melts: https://github.com/apache/spark/pull/36150/files#diff-291963840974b7fb7779e97922d3288cbbd6e2c3c029474e2bc9e7cb1d3a26e6R199-R230

Reply:
Ah sorry, I misunderstood the melt semantic.

What should be the expectation here? Do we allow auto-generated alias names? It seems safer to me to require id and value columns to either be attributes, or have explicit alias names.

Reply:
With auto-generated you mean names like these: `Alias(expr, toPrettySQL(expr))()`?

If the user is not happy with the name of their expression, then `alias` / `name` / `as` is there:

```
    val melted = ds.melt(
        Array(($"id" * 10).as("primary"), $"str1".as("secondary")),
        Array(($"int1" + $"long1").as("sum"), length($"str2").as("len")),
        variableColumnName = "var",
        valueColumnName = "val")
```

Are there other methods on `Dataset` that have a similar requirement?

Reply:
Very few operators require named expressions (AFAIK only Project and Aggregate), and they all allow auto-generated alias names. I don't have a strong opinion here, but it's always easier to support auto-generated alias names later, than removing it later.